lib/Serialization/Deserialization.cpp

```diff
@@ -166,10 +167,27 @@ static void skipRecord(llvm::BitstreamCursor &cursor, unsigned recordKind) {
 }
 
 void ModuleFile::fatal(llvm::Error error) const {
+  Core->fatal(diagnoseFatal(std::move(error)));
+}
+
+llvm::Error ModuleFile::diagnoseFatal(llvm::Error error) const {
```

Comment:
I am worried by the wide reaching effect of this PR. Recovering from previously fatal errors will likely make investigating actual fatal errors harder. The existing recovery logic has already been hiding many compilation breaking bugs, uncovering the bugs recovered from is much harder. Even when compilation ultimately fails, if we pass up an error instead of aborting we lose the stacktrace pointing to the precise point of failure.

We could print the libunwind stacktrace here but we'd still lose the support of crash tracing tools. Limiting the new recovery logic is likely the best way to go.